server/nkey.go

```diff
@@ -0,0 +1,39 @@
+// Copyright 2018 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"encoding/base64"
+)
+
+// Raw length of the nonce challenge
+const (
+	nonceRawLen = 16
+	nonceLen    = 22 // base64.RawURLEncoding.EncodedLen(nonceRawLen)
+)
+
+// nonceRequired tells us if we should send a nonce.
+// Assumes server lock is held
+func (s *Server) nonceRequired() bool {
+	return len(s.opts.Nkeys) > 0
+}
+
+// Generate a nonce for INFO challenge.
+// Assumes server lock is held
+func (s *Server) generateNonce(n []byte) {
+	var raw [nonceRawLen]byte
+	data := raw[:]
+	s.prand.Read(data)
+	base64.RawURLEncoding.Encode(n, data)
```

Comment:
Was wondering if we could simply use base32 instead (since we use that for encoding of nkeys).

Reply:
We could, would make it very big. Base32 is because we want keys themselves to be readable and selectable (double or triple click) on most OSes. The nonce and sign just need to be some form of printable hence the base64 version instead.